### ISSUE TITLE
What should .papi file format be?

### ISSUE THREAD

**Issue by Zamanke** [alleyinteractive/apple-news#875]
after the update to v2.20, there is nowhere to put in API channel/key/secret; instead I can either upload a .papi file or manually write out the contents. My current format looks like this:

`endpoint: https://news-api.apple.com/`
`channel_id: <channel_id_without_karats>`
`key: <key_without_karats>`
`secret: <secret_without_karats>`

However when I try to publish an article I get a WRONG_SIGNATURE error, which I assume has something to do with the above file? how can i fix this?

**Comment by kevinfodness** [comment]
Hi @Zamanke —

Thanks for the issue report. I can't reproduce the issue myself using my own `.papi` file.

Apple has instructions on how to connect to the News API using a `.papi` file, which are linked from the settings page, and go here: https://developer.apple.com/documentation/apple_news/apple_news_api/getting_ready_to_publish_and_manage_your_articles

Click on the link for downloading the Apple News API Client Utility. In that ZIP file will be a set of instructions on how the `.papi` file should be formatted (which roughly matches the format you are using) and contains instructions on how to verify your information by connecting to the API via the CLI utility. If you are able to successfully connect to the API using the CLI utility, then your `.papi` file is formatted correctly.

If you are able to connect via the CLI but still can't get the plugin to work correctly, could you please follow these debugging steps and report back:

1. What browser and operating system are you using? We tested using a range of browsers and operating systems, but it's possible that there was a combination that we didn't test that's causing an issue.
2. What's happening behind the scenes is the file upload box reads the contents of the file using JavaScript and puts the file contents into the textarea, then parses it using a regular expression, copies over the relevant values to hidden fields, and then the save proceeds normally. Can you verify that, once you select your `.papi` file, when you click on the `Enter the contents of your .papi file manually` link, that the contents appear in the textarea? If the textarea is blank, that would indicate a problem either with reading your file or with interacting with the form programmatically.
3. If the contents of the file appear correctly in the textarea, do you see any errors in the browser console? In Chrome, you would access the console by View > Developer > JavaScript console. Any errors would appear in red, and any warnings would appear in yellow.
4. Finally, we need to ensure that the hidden fields are being populated with the correct information. When you select your `.papi` file, or when you paste the contents into the textarea, you should be able to open the inspector for the page (I usually do this by right clicking and selecting Inspect Element in Chrome) and then digging down to the hidden elements and inspecting their values. The path to the form is: `body > div#wpwrap > div#wpcontent > div#wpbody > div#wpbody-content > div.apple-news-settings > form#apple-news-settings-form`. Within the form, if you expand the first `table > tbody > tr > td > fieldset`, the first label is the file upload, the second label is the textarea, and the final three labels are the channel, key, and secret. You should be able to see the value attribute for each, and the value attribute should match what is in your `.papi` file. If it does not, then that means that there was a problem copying those values from what is in the textarea. If it does, then the issue is with saving the field values to options.
5. If the field values match but aren't saving properly, you should check your PHP error logs on your server to determine if something is breaking down there.

**Comment by kevinfodness** [comment]
I might have a lead on this - is it possible that one of the fields (channel ID, key, secret) contains a `/`?

**Comment by Zamanke** [comment]
Hi @kevinfodness 
Thanks for the detailed reply! I am getting Javascript error messages saying `WebSocket connection to 'wss://public-api.wordpress.com/pinghub/wpcom/me/newest-note-data' failed:` and referencing `?v=2.0:formatted: 187`

In addition, looking at the hidden fields as you suggested, the value attribute for the key does not match what's in the .papi file. I'm not sure if it's significant, but the value attribute only matches the part of the key preceding a '+'

Any ideas on how to fix this?

**Comment by kevinfodness** [comment]
Thanks, that's really helpful. The web socket issue is unrelated, but the `+` tells me that it's the same issue I'm seeing. I'll work on getting a fix out for this soon.

**Comment by macstainless** [comment]
I'm running into this issue as well. My secret key has "+" and "=" in it and I'm getting the same error. Is there an ETA on when this fix may be deployed? It's completely stopping me from getting set up & publishing to Apple News. Thank you!

**Comment by Zamanke** [comment]
If it's not too much trouble, you can simply keep asking Apple for a new secret until it gives you one without any `+` signs. `=` signs don't seem to cause any trouble

**Comment by macstainless** [comment]
> If it's not too much trouble, you can simply keep asking Apple for a new secret until it gives you one without any `+` signs. `=` signs don't seem to cause any trouble

Funny, I did that before commenting and all the keys I got had other symbols besides ```=``` in it. Go figure, I tried again just now and I got a key with just ```=``` and I'm good. I do still hope this fix goes in, but more important the documentation for the Settings page on here needs to be updated. It's showing the old version and it took me a long while to figure out what I needed to do, especially with creating the .papi file.

Thank you for replying so quick.
